frontend/backend: modify transaction loading flow and add fiat value at time of transaction

Beerosagos opened the pull request:
/transactions endpoint modified to load only data displayed in transaction frontend list

A new /transaction endpoint created to load all available data for specific transaction in order to populate the detailed transaction view in frontend

FiatConversion component modified to use backend-calculated fiat amount conversions when available

Added code to show fiat value at time of transaction in detailed transaction view

benma commented:
@Beerosagos cool!

please modify the commit message to follow https://github.com/digitalbitbox/bitbox-wallet-app/blob/master/CONTRIBUTING.md#git-commits, especially 

> The goal is that reading the commit message on any line using git blame explains why the line exists.

I.e. why over what.

e.g. why was the fiatconversion component modified?

also please try to split changes into atomic commits in the future, e.g. the refactor of the transactions loading can be a first commit, followed by adding the new historical fiat value entry in another commit, so the changes are easier to review.

benma commented on backend/coins/btc/handlers/handlers.go:
```diff
@@ -101,11 +102,25 @@ func (handlers *Handlers) formatAmountAsJSON(amount coin.Amount, isFee bool) For
 	}
 }
 
+func (handlers *Handlers) formatAmountAtTimeAsJSON(amount coin.Amount, isFee bool, timeStamp *time.Time) FormattedAmount {
```
remove the isFee param, as it is always `false`?

or is the plan to also add the fee fiat at the time of transaction?

Beerosagos replied:
Good point! I don't think we are planning to add historical fiat value for the fee. I'll just remove it.

benma commented on frontends/web/src/api/account.ts:
```diff
@@ -184,6 +185,9 @@ export const getTransactionList = (code: AccountCode): Promise<ITransaction[]> =
   return apiGet(`account/${code}/transactions`);
 };
 
+export const getTransaction = (code: AccountCode, id: ITransaction['txID']): Promise<ITransaction> => {
```
should be `internalID` not `txID`?

interesting trick, didn't know could could reference a type in an interface like that.

Beerosagos replied:
You're right, I'll change it! Thx

benma commented on backend/coins/btc/handlers/handlers.go:
```diff
@@ -53,6 +53,7 @@ func NewHandlers(
 	handleFunc("/init", handlers.postInit).Methods("POST")
 	handleFunc("/status", handlers.getAccountStatus).Methods("GET")
 	handleFunc("/transactions", handlers.ensureAccountInitialized(handlers.getAccountTransactions)).Methods("GET")
+	handleFunc("/transaction", handlers.ensureAccountInitialized(handlers.getTransaction)).Methods("GET")
```
consistency: add or remove `Account` in both function names

Beerosagos replied:
Didn't use `getAccountTransaction` because in my mind: 
`getAccountTransactions` -> retrieve all the transaction of a certain account
`getTransaction` -> retrieve a specific transaction (by ID)

I wouldn't remove `Account` from  `getAccountTransactions`, but I see no problems in making `getAccountTransaction`

benma commented on backend/coins/btc/handlers/handlers.go:
```diff
@@ -171,11 +187,68 @@ func (handlers *Handlers) getAccountTransactions(_ *http.Request) (interface{},
+	return result, nil
+}
+
+func (handlers *Handlers) getTransaction(r *http.Request) (interface{}, error) {
```
a lot in this function is now duplicated with the contents of the other function, `getAccountTransactions`.

I think it's worthwhile to try to refactor it so that the duplicated code is removed, i.e. a helper function that both functions call or similar.

Beerosagos replied:
Agree :+1:

Beerosagos commented:
Thanks @benma. I updated the code to address your observations. 

I actually made two different commits, but then I squashed them before the PR. Next time I'll avoid that. 
I am not sure how to improve the commit message without using a very long first line. Do you have any suggestion about that?

benma commented:
>  I am not sure how to improve the commit message without using a very long first line. Do you have any suggestion about that?

The commit title should be a short summary, <=68 chars, e.g. `backend: refactor /transactions and add historical fiat value`

The commit message body can be as long as needed, including all the details as to why some change happened (e.g why the fiatconversion component needed to be modified this way).

Beerosagos commented:
I updated the commit massage, hope it is better now. Thanks for the help!

benma commented:
> FiatConversion component modified to use high-precision backend-calculated fiat amount
conversions when available, instead of doing the math in the frontend with lower precision

Is this the only reason? because the backend doesn't use high precision. it uses floating point arith, see 

https://github.com/digitalbitbox/bitbox-wallet-app/pull/1724/files#diff-7a00117689d4febac66fe55f84fd7cb00c99873d547c480c09d1416b57e33647R49

it seems the actual reason is because the historical rates are only available in the backend and not the frontend.

Beerosagos commented on frontends/web/src/components/rates/rates.tsx:
```diff
@@ -143,13 +138,19 @@ function Conversion({
+    if (rates && rates[coin]) {
+      // TODO This solution is not the best: it is better to have the backend calculating the value,
+      // with better precision, using coins/convertToFiat endpoint
+      formattedValue = formatCurrency(rates[coin][active] * Number(amount.amount), active);
```
> > FiatConversion component modified to use high-precision backend-calculated fiat amount
> > conversions when available, instead of doing the math in the frontend with lower precision
> 
> Is this the only reason? because the backend doesn't use high precision. it uses floating point arith, see
> 
> https://github.com/digitalbitbox/bitbox-wallet-app/pull/1724/files#diff-7a00117689d4febac66fe55f84fd7cb00c99873d547c480c09d1416b57e33647R49
> 
> it seems the actual reason is because the historical rates are only available in the backend and not the frontend.

I think you are right, both frontend and backend use 64bit floating point for the math, and we only have historical rates in the backend. I talked with @thisconnect when I started working on this issue and I understood that it was better to avoid coin conversions in frontend, but maybe I remember wrong and the reason was not the precision?

thisconnect replied:
BBApp usually never calculates in the frontend and I assumed backend does more precise calculation sorry :) but now it could right?

Beerosagos replied:
Oh, ok! In that case I would just remove the comment at https://github.com/Beerosagos/bitbox-wallet-app/blob/transaction-detail/frontends/web/src/components/rates/rates.tsx#L150-L151

Ok?

thisconnect replied:
no strong opinion, but the comment is still valid or not?

Beerosagos replied:
I think that, until the precisions in the backend and in frontend are the same, this comment loses its sense.

Beerosagos commented on backend/coins/btc/handlers/handlers.go:
```diff
@@ -139,43 +155,46 @@ func (handlers *Handlers) ensureAccountInitialized(h func(*http.Request) (interf
-			Addresses: addresses,
-			Note:      handlers.account.TxNote(txInfo.InternalID),
-		}
+func (handlers *Handlers) getTxInfoJSON(txInfo *accounts.TransactionData, detail bool) Transaction {
```
ok!

Beerosagos commented on frontends/web/src/locales/it/app.json:
```diff
@@ -1174,6 +1174,7 @@
       "date": "Data",
       "fiat": "Fiat",
       "fiatAmount": "Importo fiat",
+      "fiatAtTime": "Fiat al tempo della transazione",
```
Ok, I'll remove it for clarity. I didn't find the doc explaining the locize part while updating the code. Now that I read it, it seems quite clear to me.

Beerosagos commented on backend/coins/btc/handlers/handlers.go:
```diff
@@ -189,11 +208,43 @@ func (handlers *Handlers) getAccountTransactions(_ *http.Request) (interface{},
+
+		return handlers.getTxInfoJSON(txInfo, true), nil
+	}
+	return txInfoJSON, nil
```
Done :+1:

Beerosagos commented on backend/coins/btc/handlers/handlers.go:
```diff
@@ -189,11 +208,43 @@ func (handlers *Handlers) getAccountTransactions(_ *http.Request) (interface{},
+func (handlers *Handlers) getAccountTransaction(r *http.Request) (interface{}, error) {
+	internalID := r.URL.Query().Get("id")
+	if internalID == "" {
+		return Transaction{}, errp.New("Transaction ID is missing.")
```
ok, perfect!

Beerosagos commented on frontends/web/src/api/account.ts:
```diff
@@ -184,6 +185,9 @@ export const getTransactionList = (code: AccountCode): Promise<ITransaction[]> =
   return apiGet(`account/${code}/transactions`);
 };
 
+export const getTransaction = (code: AccountCode, id: ITransaction['internalID']): Promise<ITransaction> => {
```
Done :+1:

Beerosagos commented:
Pushed an updated commit

benma reviewed (approved):
awesome!

please update the CHANGELOG :)
